### PR TITLE
HGI-6568 / Add advocacy_campaigns/outreaches

### PR DESCRIPTION
**Pull request by arilton**

### CODE REVIEW

File: target_actionnetwork/client.py [hunk 1]
```diff
@@ -43,8 +43,8 @@ def _request(
     
     @property
     def base_url(self) -> str:
-        return self.config.get("uri")
+        return "https://actionnetwork.org/api/v2/"
```
Comment (ghost) [inline target_actionnetwork/client.py:46]: Why did you change it?

File: target_actionnetwork/target.py [hunk 2]
```diff
@@ -1,26 +1,23 @@
...
-]
-
 class TargetActionNetwork(Target, TargetHotglue):
     """Sample target for ActionNetwork."""
 
     name = "target-actionnetwork"
 
     MAX_PARALLELISM = 1
 
+    SINK_TYPES = [
```
Comment (ghost) [inline target_actionnetwork/target.py:22]: Any particular reason why you made this change?

File: target_actionnetwork/sinks.py [hunk 3]
```diff
@@ -96,10 +159,28 @@ def upsert_record(self, record: dict, context: dict):
             if record.get("error"):
                 raise Exception(record.get("error"))
 
+            lists = None
+            if "lists" in record:
+                lists = record.pop("lists")
```
Comment (ghost) [inline target_actionnetwork/sinks.py]:
```suggestion
           lists = record.pop("lists",None)
```

File: target_actionnetwork/sinks.py [hunk 4]
```diff
@@ -38,12 +101,14 @@ def preprocess_record(self, record: dict, context: dict) -> dict:
 
         #One of ['subscribed', 'unsubscribed', 'bouncing', 'previou' bounce’, 'spa' complaint’, or 'previou' spam complaint’]
         status = None
-        if (subscribe_status := record.get("subscribe_status")) in ['subscribed', 'unsubscribed', 'bouncing', 'previou bounce', 'spa complaint', 'previou spam complaint']:
+        subscribe_status = record.get("subscribe_status")
+        if subscribe_status and subscribe_status in ['subscribed', 'unsubscribed', 'bouncing', 'previou bounce', 'spa complaint', 'previou spam complaint']:
             status = subscribe_status
         elif record.get("unsubscribed"):
```
Comment (ghost) [inline target_actionnetwork/sinks.py:105]: Did you came across any scenario that it makes sense?

File: target_actionnetwork/sinks.py [hunk 5]
```diff
@@ -38,12 +101,14 @@ def preprocess_record(self, record: dict, context: dict) -> dict:
 
         #One of ['subscribed', 'unsubscribed', 'bouncing', 'previou' bounce’, 'spa' complaint’, or 'previou' spam complaint’]
         status = None
-        if (subscribe_status := record.get("subscribe_status")) in ['subscribed', 'unsubscribed', 'bouncing', 'previou bounce', 'spa complaint', 'previou spam complaint']:
+        subscribe_status = record.get("subscribe_status")
+        if subscribe_status and subscribe_status in ['subscribed', 'unsubscribed', 'bouncing', 'previou bounce', 'spa complaint', 'previou spam complaint']:
```
Comment (ghost) [inline target_actionnetwork/sinks.py]:
```suggestion
        if subscribe_status in ['subscribed', 'unsubscribed', 'bouncing', 'previou bounce', 'spa complaint', 'previou spam complaint']:
```

File: target_actionnetwork/sinks.py [hunk 6]
```diff
@@ -1,15 +1,74 @@
...
+    advocacy_campaigns = {}
+
+    def __init__(self, target, stream_name, schema, key_properties):
+        super().__init__(target, stream_name, schema, key_properties)
+        self.get_advocacy_campaigns()
+
+    def get_advocacy_campaigns(self):
+        response = self._request("GET", "advocacy_campaigns")
+        res_json = response.json()
+        total_pages = res_json.get("total_pages")
```
Comment (ghost) [inline target_actionnetwork/sinks.py:28]: You could extract this session to another function called `_fetch_advocacy_campaign_pages `. Where you would return the `total_pages`

File: target_actionnetwork/sinks.py [hunk 7]
```diff
@@ -1,15 +1,74 @@
...
+        current_page = 0
+        while current_page < total_pages:
+            current_page += 1
```
Comment (ghost) [inline target_actionnetwork/sinks.py]: You could exchange that for a `for` loop to make it more clear and safe

File: target_actionnetwork/sinks.py [hunk 8]
```diff
@@ -1,15 +1,74 @@
...
+            response = self._request("GET", "advocacy_campaigns", {"page": current_page})
+            res_json = response.json()
+            advocacy_campaigns = res_json["_embedded"]["osdi:advocacy_campaigns"]
+            for adv_camp in advocacy_campaigns:
+                if adv_camp["title"] in self.advocacy_campaigns:
+                    self.logger.warning(f"The campaign (title={adv_camp['title']}, id={adv_camp['identifiers']}) already exists, and it is mapped to id={self.advocacy_campaigns[adv_camp['title']]}")
+                self.advocacy_campaigns[adv_camp["title"]] = adv_camp["identifiers"][0].split(':')[1]
```
Comment (ghost) [inline target_actionnetwork/sinks.py:36]: You could extract this session to another method to control better your context

**Review by ghost — APPROVED** [review]
Excellent work @arilton . Just left some comments and questions, but they are not blockers.

File: target_actionnetwork/client.py (hunk 1 above)
Comment (arilton) [reply target_actionnetwork/client.py:46]: i first put here and it was one of my first tickets here in hotglue. after work more weeks here i realize that our pattern is to have it usually here hardcoded, the exceptions is to put in the config. what it is not this case

File: target_actionnetwork/target.py (hunk 2 above)
Comment (arilton) [reply target_actionnetwork/target.py:22]: no, put it back

File: target_actionnetwork/sinks.py (hunk 6 above)
Comment (arilton) [reply target_actionnetwork/sinks.py:28]: it could be confusing to have a `get_advocacy_campaigns` and `_fetch_advocacy_campaign_pages`, so i prefer to not do it

File: target_actionnetwork/sinks.py (hunk 8 above)
Comment (arilton) [reply target_actionnetwork/sinks.py:36]: i need th current page, so i will still need this context and not achieve the goal of add another method

**Review by ghost — APPROVED** [review]
